text/0122-sdk-hub-scope-merge.md

```diff
@@ -0,0 +1,475 @@
+
+1. To ensure the experience is somewhat consistent across SDKs, and concepts can be ported.
+2. To handle isomorphic cases, e.g. in JS you have Server and Client code, and it should be as portable as possible.
+3. To prepare for the future - even if e.g. multi-execution-context is not important in an SDK right now, it may become so in the future.
+
+# Proposed Solution
+
+Hubs will be removed as a concept for SDK users.
+Instead, from a users perspective only Scopes remain, which will become the singular entity to hold context data etc.
+
+Scopes will be _similar_ to how they work today, but not entirely the same.
+Scopes can have data (e.g. tags, user, ...) added to them the same way you can do today.
+This RFC _does not_ aim to change any of the data that is kept on the scope and is applied to events.
+
+The following APIs will be deprecated:
+
+- `getCurrentHub()`
+- `configureScope()` (instead just get the scope and set on it directly)
+- Any APIs currently on the Hub only:
+  - `hub.pushScope()`
+  - `hub.popScope()`
+  - `hub.isOlderThan()`
+  - `hub.bindClient()`
+  - `hub.getStack()`
+  - `hub.getStackTop()`
+  - `hub.run()` (use `withScope()` instead)
+
+Instead, we will introduce some new APIs:
+
+```ts
+// get the current scope. replacement for `getCurrentHub().getScope()`
+export function getCurrentScope(): Scope;
+
+// get the current client. May return a NOOP client. Replacement for `getCurrentHub().getClient()`.
+export function getClient(): Client;
+
+// make a scope the current scope. Replacement for `makeMain(hub)`
+export function setCurrentScope(scope: Scope): void;
+
+// get the current global scope
+export function getGlobalScope(): Scope;
+
+// get the curren isolation scope
+export function getIsolationScope(): Scope;
+
+// similar to `withScope`, but defines an isolation scope
+export function withIsolationScope(
+  callback: (scope: Scope) => unknown
+): unknown;
+```
+
+The following APIs already exist but will behave differently:
+
+- `withScope()` will still work, but it will actually fork an execution context. So this will roughly do the same as doing `hub.run()` today in languages that have that, which forks an execution context.
+
+APIs that are currently on the hub should instead be called directly on the scope (e.g. `scope.captureException()` etc.), or via a global method (e.g. `Sentry.captureException()`).
+
+The current scope may be kept similar to how we currently keep the current hub, but this is SDK specific and not part of this RFC.
+
+You can find some examples for how the APIs can/should be used [here](https://gist.github.com/mitsuhiko/1bc78d04ea7d08e5b50d27e42676db80).
+
+## Clients
+
+Instad of picking the client from the current hub, going forward a client should always be fetched via `Sentry.getClient()`. 
+This should _always_ return a client. If `Sentry.init()` was not called yet, it should return a Noop Client that does nothing.
```

Comment:
> Instad of picking the client from the current hub, going forward a client should always be fetched via `Sentry.getClient()`.

This feels odd. If SDK users are only using one client (`Sentry.getClient()` implies a static/singleton reference to a Sentry Client) then this could work. It might be easier to make the Scope Stack a property of the client though and use typical DI techniques to get a reference to the Client... avoiding all the problems with statics/singletons.

If SDK users have multiple clients, getting the client via `Sentry.getClient()` is problematic. It results in weird things like temporarily pushing a different client to the scope stack (so `Sentry.getClient()` can return the top/current client from the stack)... in the case of Global/Shared scope, this could easily lead to problems. 

For SDK users and SDK designers, it would be easier if the ScopeStack was a property of the client. **_If_** SDK users wanted two different clients to share the same Scope, those clients could be configured/initialised with some kind of shared scope stack (like the `GlobalScope` from this RFC)... but events would always end up being sent to the correct Sentry instance/project. 

I'm not sure I understand why/when anyone would want the client to be ambient context (in the same way that scopes are) or why anyone would maintain a stack of clients (except as a workaround to our SDK design - namely that we retrieve the client reference via a static/singleton reference).

Reply:
This should just work like other things on static API that interact with the scope by creating a sort of view on top of global, isolation and current scope, then handing back whatever wins according to logic that should be detailed somewhere in this RFC. There shouldn't be a need to change the stack and stack should in theory go away with the proposed changes here.